Fix #531: Docker changed their cgroups format

jbfavre opened the pull request:
Take into account new cgroup format for docker.
Tested with docker 1.6.2 on Debian sid with kernel 4.0.5

chef-supermarket commented:
Hi. I am an automated pull request bot named Curry. There are commits in this pull request whose authors are not yet authorized to contribute to Chef Software, Inc. projects or are using a non-GitHub verified email address. To become authorized to contribute, you will need to sign the Contributor License Agreement (CLA) as an individual or on behalf of your company. [You can read more on Chef's blog.](https://www.chef.io/blog/2014/06/23/changes-to-the-contributor-license-agreement-process)

## GitHub Users Who Are Not Authorized To Contribute

The following GitHub users do not appear to have signed a CLA:

* @jbfavre

[Please sign the CLA here.](https://supermarket.chef.io/become-a-contributor)

chef-supermarket commented:
Hi. Your friendly Curry bot here. Just letting you know that all commit authors have become authorized to contribute. I have added the "Signed CLA" label to this issue so it can easily be found in the future.

mcquin commented on lib/ohai/plugins/linux/virtualization.rb:
```diff
@@ -186,7 +186,8 @@ def docker_exists?
     if File.exists?("/proc/self/cgroup")
-      if File.read("/proc/self/cgroup") =~ %r{^\d+:[^:]+:/(lxc|docker)/.+$}
+      if File.read("/proc/self/cgroup") =~ %r{^\d+:[^:]+:/(lxc|docker)/.+$} ||
+         File.read("/proc/self/cgroup") =~ %r{^\d+:[^:]+:/[^/]+/(lxc|docker)-.+$}
```
Avoid reading twice by assigning `File.read("/proc/self/cgroup")` to a variable.

mcquin commented:
@jbfavre Can you add a spec for the new cgroup format [here](https://github.com/chef/ohai/blob/master/spec/unit/plugins/linux/virtualization_spec.rb#L457)?

jbfavre commented:
@mcquin added required spec & also apply a fix based on your first comment

mcquin commented:
Great! Thanks!

:+1:

@chef/client-core

danielsdeleo commented on spec/unit/plugins/linux/virtualization_spec.rb:
```diff
@@ -493,6 +493,25 @@
       expect(@plugin[:virtualization][:systems][:docker]).to eq("guest")
     end
 
+    it "should set docker guest if /proc/self/cgroup exists and there are /system.slice/docker-<hexadecimal> mounts" do
```
Would be awesome to have a note about which version of Docker this applies to, either here or in the non-test code.

danielsdeleo commented:
:+1: Would like a comment about which version of docker uses which format, but I wouldn't block merging this on it.

jbfavre commented:
@danielsdeleo All I know is that it works with docker 1.6.2. Didn't had the opportunity to test against earlier release. Is it enough ?

danielsdeleo commented:
Yeah, if we don't know the exact release version of the change, that's okay. A comment saying that the new format is _the_ format as of at least 1.6.2 would come in handy in the future when we're wondering if we can stop supporting the old format.

jbfavre commented:
@danielsdeleo I added requested comment. Also rebased against master to solve conflict. Sorry for so many commits, I'm definitely not a Ruby guru :)

I also wonder if the cgroup format is due to docker or systemd. I think this change was introduced by systemd. I added systemd version as well

danielsdeleo commented:
Excellent, thank you!

thommay commented:
squashed and merged. Thanks!
